### PR TITLE
Headers Performance Boost and Interface Simplification

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
A degradation in performance has been observed from the 4.0 branch as documented in https://github.com/netty/netty/issues/3962.

Modifications:
- Simplify Headers class hierarchy.
- Restore the DefaultHeaders to be based upon DefaultHttpHeaders from 4.0.
- Make various other modifications that are causing hot spots.

Result:
Performance is now on par with 4.0.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@normanmaurer - PTAL. Basically everything in the `HttpHeaders` / `DefaultHttpHeaders` is `deprecated`. Most `deprecated` items have alternatives to use, but some describe that future releases will be changing this method (rather than bringing *everything* back, and some type signatures in master conflict).

File: codec-http/src/main/java/io/netty/handler/codec/http/EmptyHttpHeaders.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2014 The Netty Project
+ *
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/EmptyHttpHeaders.java]: 2015 ?

File: codec-http/src/test/java/io/netty/handler/codec/http/HttpHeaderUtilTest.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2013 The Netty Project
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpHeaderUtilTest.java]: 2015

File: codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTestUtils.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    public enum HeaderValue {
+        UNKNOWN("unknown", 0),
+        ONE("one", 1),
+        TWO("two", 2),
+        THREE("three", 3),
+        FOUR("four", 4),
+        FIVE("five", 5),
+        SIX_QUOTED("six,", 6),
+        SEVEN_QUOTED("seven; , GMT", 7),
+        EIGHT("eight", 8);
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTestUtils.java]: I think we usually use CamelCase

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java [hunk 4]
```diff
@@ -211,54 +241,48 @@ public int size() {
...
     @Override
     public boolean isEmpty() {
-        return map.isEmpty();
+        return head == head.after;
     }
 
     @Override
     public Set<T> names() {
-        return unmodifiableSet(map.keySet());
+        Set<T> names = new LinkedHashSet<T>(size());
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: I guess we could first check if its empty and if so just use Collections.emptySet() ?

File: codec-http/src/main/java/io/netty/handler/codec/http/EmptyHttpHeaders.java (hunk 1 above)
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/EmptyHttpHeaders.java]: yip. fixed.

File: codec-http/src/test/java/io/netty/handler/codec/http/HttpHeaderUtilTest.java (hunk 2 above)
Comment (Scottmitch) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpHeaderUtilTest.java]: done

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java (hunk 4 above)
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: Sure. Done.

File: codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTestUtils.java (hunk 3 above)
Comment (Scottmitch) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTestUtils.java]: Done

**Comment by Scottmitch** [comment]
@normanmaurer - Addressed comments. My thought is to get this PR to an acceptable state, cherry-pick it, and then discuss the deprecation changes with @trustin and @nmittler which may result in a followup PR. Sound good?

**Comment by trustin** [comment]
SGTM!

**Comment by normanmaurer** [comment]
Ship it

**Comment by Scottmitch** [comment]
Cherry-picked 4.1 (ba6ce54).

Thanks for review all!

**Pull request closed without merging**